adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpContext.java

```diff
@@ -30,22 +29,22 @@
     private final Message message;
     private ResourceIdentifier resource;
     private final Device authenticatedDevice;
-    private final Target receiverTarget;
 
     AmqpContext(final ProtonDelivery delivery, final Message message, final Target receiverTarget, final Device authenticatedDevice) {
         this.delivery = delivery;
         this.message = message;
         this.authenticatedDevice = authenticatedDevice;
-        this.receiverTarget = receiverTarget;
         if (receiverTarget == null || receiverTarget.getAddress() == null) {
             // link possibly opened by an "authenticated" gateway component via anonymous relay
-            validateMessageProperty(message.getProperties().getTo()).compose(ok -> {
-                resource = ResourceIdentifier.fromString(message.getProperties().getTo());
-                return Future.succeededFuture();
-            }).recover(t -> {
-               handleFailure((ServiceInvocationException) t);
-               return Future.failedFuture(t);
-            });
+            try {
+                validateMessageAddress(message.getAddress());
+                this.resource = ResourceIdentifier.fromString(message.getAddress());
+            } catch (ServiceInvocationException sie) {
+                handleFailure(sie);
+            }
+
+        } else if (isDeviceAuthenticated()) {
+            this.resource = ResourceIdentifier.from(receiverTarget.getAddress(), authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
```

Comment:
is there a reason why the target address can only contain the endpoint name in this case?

Reply:
yes, for authenticated devices(non-gateway), the `endpointName` is specified in the address